plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -98,6 +105,139 @@ instance Pretty ChainIndexTxOut where
+  parseJSON "ShelleyEraInCardanoMode" = pure C.ShelleyEraInCardanoMode
+  parseJSON invalid =
+      prependFailure "parsing 'EraInMode ShelleyEra CardanoMode' failed, "
+                     (typeMismatch "ShelleyEraInCardanoMode" invalid)
+
+-- TODO Move to cardano-api
+instance FromJSON (C.EraInMode C.AllegraEra C.CardanoMode) where
+  parseJSON "AllegraEraInCardanoMode" = pure C.AllegraEraInCardanoMode
+  parseJSON invalid =
+      prependFailure "parsing 'EraInMode AllegraEra CardanoMode' failed, "
+                     (typeMismatch "AllegraEraInCardanoMode" invalid)
+
+-- TODO Move to cardano-api
+instance FromJSON (C.EraInMode C.MaryEra C.CardanoMode) where
+  parseJSON "MaryEraInCardanoMode" = pure C.MaryEraInCardanoMode
+  parseJSON invalid =
+      prependFailure "parsing 'EraInMode MaryEra CardanoMode' failed, "
+                     (typeMismatch "MaryEraInCardanoMode" invalid)
+
+-- TODO Move to cardano-api
+instance FromJSON (C.EraInMode C.AlonzoEra C.CardanoMode) where
+  parseJSON "AlonzoEraInCardanoMode" = pure C.AlonzoEraInCardanoMode
+  parseJSON invalid =
+      prependFailure "parsing 'EraInMode AlonzoEra CardanoMode' failed, "
+                     (typeMismatch "AlonzoEraInCardanoMode" invalid)
+
+-- TODO Move to cardano-api
+instance ToJSON (C.EraInMode era mode) where
+  toJSON C.ByronEraInByronMode     = "ByronEraInByronMode"
+  toJSON C.ShelleyEraInShelleyMode = "ShelleyEraInShelleyMode"
+  toJSON C.ByronEraInCardanoMode   = "ByronEraInCardanoMode"
+  toJSON C.ShelleyEraInCardanoMode = "ShelleyEraInCardanoMode"
+  toJSON C.AllegraEraInCardanoMode = "AllegraEraInCardanoMode"
+  toJSON C.MaryEraInCardanoMode    = "MaryEraInCardanoMode"
+  toJSON C.AlonzoEraInCardanoMode  = "AlonzoEraInCardanoMode"
+
+data SomeCardanoApiTx where
+  SomeTx :: C.IsCardanoEra era => C.Tx era -> C.EraInMode era C.CardanoMode -> SomeCardanoApiTx
+
+instance Eq SomeCardanoApiTx where
+  (SomeTx tx1 C.ByronEraInCardanoMode) == (SomeTx tx2 C.ByronEraInCardanoMode)     = tx1 == tx2
+  (SomeTx tx1 C.ShelleyEraInCardanoMode) == (SomeTx tx2 C.ShelleyEraInCardanoMode) = tx1 == tx2
+  (SomeTx tx1 C.AllegraEraInCardanoMode) == (SomeTx tx2 C.AllegraEraInCardanoMode) = tx1 == tx2
+  (SomeTx tx1 C.MaryEraInCardanoMode) == (SomeTx tx2 C.MaryEraInCardanoMode)       = tx1 == tx2
+  (SomeTx tx1 C.AlonzoEraInCardanoMode) == (SomeTx tx2 C.AlonzoEraInCardanoMode)   = tx1 == tx2
+  _ == _                                                                           = False
+
+deriving instance Show SomeCardanoApiTx
+
+instance ToJSON SomeCardanoApiTx where
+  toJSON (SomeTx tx eraInMode) =
+    object [ "tx" .= C.serialiseToTextEnvelope Nothing tx
+           , "eraInMode" .= eraInMode
+           ]
+
+-- | Converting 'SomeCardanoApiTx' to JSON.
+--
+-- If the "tx" field is from an unknown era, the JSON parser print the error
+-- message corresponding to the last 'Alternative' in the 'parseJSON'
+-- implementation.
```

Comment:
You could add a final parser that always fails, with a better message 

`parseEraInCardanoModeFail _ = fail "Unable to parse 'eraInMode'"`

Reply:
Good idea! Done.